gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcCallContext.java

```diff
@@ -170,12 +180,66 @@ public GrpcCallContext withStreamIdleTimeout(@Nullable Duration streamIdleTimeou
     }
 
     return new GrpcCallContext(
-        channel, callOptions, streamWaitTimeout, streamIdleTimeout, channelAffinity);
+        channel, callOptions, streamWaitTimeout, streamIdleTimeout, channelAffinity, extraHeaders);
   }
 
+  @BetaApi("The surface for channel affinity is not stable yet and may change in the future.")
   public GrpcCallContext withChannelAffinity(@Nullable Integer affinity) {
     return new GrpcCallContext(
-        channel, callOptions, streamWaitTimeout, streamIdleTimeout, affinity);
+        channel, callOptions, streamWaitTimeout, streamIdleTimeout, affinity, extraHeaders);
+  }
+
+  @BetaApi("The surface for extra headers is not stable yet and may change in the future.")
+  @Override
+  public GrpcCallContext withExtraHeaders(Map<String, List<String>> extraHeaders) {
+    Preconditions.checkNotNull(extraHeaders);
+    ImmutableMap<String, List<String>> newExtraHeaders =
+        mergeExtraHeaders(this.extraHeaders, extraHeaders);
+    return new GrpcCallContext(
+        channel,
+        callOptions,
+        streamWaitTimeout,
+        streamIdleTimeout,
+        channelAffinity,
+        newExtraHeaders);
+  }
+
+  @BetaApi("The surface for extra headers is not stable yet and may change in the future.")
+  @Override
+  public GrpcCallContext withEmptyExtraHeaders() {
+    return new GrpcCallContext(
+        channel,
+        callOptions,
+        streamWaitTimeout,
+        streamIdleTimeout,
+        channelAffinity,
+        ImmutableMap.<String, List<String>>of());
+  }
+
+  private static ImmutableMap<String, List<String>> mergeExtraHeaders(
+      Map<String, List<String>> oldExtraHeaders, Map<String, List<String>> newExtraHeaders) {
+    ImmutableMap.Builder<String, List<String>> builder =
+        ImmutableMap.<String, List<String>>builder();
+
+    for (Map.Entry<String, List<String>> oldHeader : oldExtraHeaders.entrySet()) {
+      String key = oldHeader.getKey();
+      if (newExtraHeaders.containsKey(key)) {
+        builder.put(
+            key,
+            ImmutableList.<String>builder()
+                .addAll(oldHeader.getValue())
+                .addAll(newExtraHeaders.get(key))
+                .build());
+      } else {
+        builder.put(oldHeader);
```

Comment:
I believe you still need to make a deep copy of the map entry, since the `List<String>` provided as a value in `oldExtraHeaders` and `newExtraHeaders` is not guaranteed to be immutable.